component/s3/file/s3.go

```diff
@@ -0,0 +1,213 @@
+	opts       component.Options
+	args       Arguments
+	health     *component.Health
+	watcher    *watcher
+	updateChan chan []byte
+	errorChan  chan error
+	content    string
+	cancel     context.CancelFunc
+}
+
+var (
+	_ component.Component       = (*S3)(nil)
+	_ component.HealthComponent = (*S3)(nil)
+)
+
+// New initializes the s3 component
+func New(o component.Options, args Arguments) (*S3, error) {
+	s3cfg, err := generateS3Config(args)
+	if err != nil {
+		return nil, err
+	}
+
+	s3Client := s3.NewFromConfig(*s3cfg, func(s3o *s3.Options) {
+		s3o.UsePathStyle = args.Options.UsePathStyle
+	})
+
+	bucket, file := getPathBucketAndFile(args.Path)
+	s := &S3{
+		opts:       o,
+		args:       args,
+		health:     &component.Health{},
+		updateChan: make(chan []byte),
+		errorChan:  make(chan error),
+	}
+
+	w, err := newWatcher(bucket, file, s.updateChan, s.errorChan, args.PollFrequency, s3Client)
+	if err != nil {
+		return nil, err
+	}
+	s.watcher = w
+	return s, nil
+}
+
+// Run activates the content handler and watcher
+func (s *S3) Run(ctx context.Context) error {
+	go s.handleContentUpdate(ctx)
+	go s.watcher.run(ctx)
+	<-ctx.Done()
+	if s.cancel != nil {
+		s.cancel()
+	}
+
+	return nil
+}
+
+// Update is called whenever the arguments have changed
+func (s *S3) Update(args component.Arguments) error {
+	newArgs := args.(Arguments)
+
+	s3cfg, err := generateS3Config(newArgs)
+	if err != nil {
+		return nil
+	}
+	s3Client := s3.NewFromConfig(*s3cfg, func(s3o *s3.Options) {
+		s3o.UsePathStyle = newArgs.Options.UsePathStyle
+	})
+
+	bucket, file := getPathBucketAndFile(newArgs.Path)
+
+	s.mut.Lock()
+	defer s.mut.Unlock()
+	s.args = newArgs
+	s.watcher.updateValues(bucket, file, newArgs.PollFrequency, s3Client)
+
+	return nil
+}
+
+// CurrentHealth returns the health of the component
+func (s *S3) CurrentHealth() component.Health {
+	s.mut.Lock()
+	defer s.mut.Unlock()
+	return *s.health
+}
+
+func generateS3Config(args Arguments) (*aws.Config, error) {
+	configOptions := make([]func(*aws_config.LoadOptions) error, 0)
+	// Override the endpoint
+	if args.Options.Endpoint != "" {
+		endFunc := aws.EndpointResolverWithOptionsFunc(func(service, region string, _ ...interface{}) (aws.Endpoint, error) {
+			return aws.Endpoint{URL: args.Options.Endpoint}, nil
+		})
+		endResolver := aws_config.WithEndpointResolverWithOptions(endFunc)
+		configOptions = append(configOptions, endResolver)
+	}
+
+	// This incredibly nested option turns off ssl
+	if args.Options.DisableSSL {
+		httpOverride := aws_config.WithHTTPClient(
+			&http.Client{
+				Transport: &http.Transport{
+					TLSClientConfig: &tls.Config{
+						InsecureSkipVerify: args.Options.DisableSSL,
+					},
+				},
+			},
+		)
+		configOptions = append(configOptions, httpOverride)
+	}
+
+	// check credentials
+	if args.Options.AccessKey != "" {
+		if args.Options.Secret == "" {
+			return nil, fmt.Errorf("if accesskey or secret are specified then the other must also be specified")
+		}
```

Comment:
Note to self: it'd be great if we could return an error for specifically this field but we can't do that right now, so the error will be shown at the component level.